cubedash/index/api.py

```diff
@@ -0,0 +1,193 @@
+from abc import ABC, abstractmethod
+from datetime import datetime, timedelta
+from typing import Generator, Iterable
+from uuid import UUID
+
+from datacube.index import Index
+from datacube.model import Dataset, MetadataType, Product, Range
+from sqlalchemy.sql import ColumnElement
+
+
+class EmptyDbError(Exception):
+    pass
+
+
+class ExplorerAbstractIndex(ABC):
```

Comment:
Eventually it would be nice to use this (and the equivalent abstract class in OWS) as a checklist of functionality to be brought into the core index API.

It would therefore be nice if all these methods could have clear docstrings.  Probably not worth holding up the 1.9 release over though.